### ISSUE TITLE
How to pass Output<T> as an Environmental Variable to a service?

### ISSUE THREAD

**Issue by volkanunsal** [pulumi/pulumi#6077]
I'd like to use the names of a few resources as environmental variables in a service that is created last. I understand I cannot use the Output object directly (since it's asynchronous), so I have taken to nesting the `apply` calls like this to extract the value of the outputs. But it's a bit messy. And I'm worried that this might cause the service to be destroyed and recreated when the value of an output up the stack changes.

Is there a better way?

![Screen Shot 2021-01-10 at 1 39 09 PM](https://user-images.githubusercontent.com/151600/104132223-cbb09f00-5349-11eb-802c-6c5a4ceca672.png)

**Comment by lukehoban** [comment]
You can write this code as:

```ts
const dockerBuild = pulumi.all([bucket.bucket, this.accessKeyId, this.secretAccessKey, this.userId, elastiCache.cacheNodes]).apply(args => {
    const [b, ak, sak, uid, cn] = args;
    return <docker.DockerBuild>{
        context: ".",
        dockerfile: "./eagle-1/Dockerfile",
        env: {
            AWS_BUCKET: b,
            AWS_ACCESS_KEY_ID: ak,
            AWS_SECRET_ACCESS_KEY: sak,
            AWS_ACCOUNT_IT: uid,
            REDIS_ADDRESS: cn[0].address,
            REDIS_PORT: cn[0].port,
        }
    };
});
const s = new awsx.ecs.FargateService("svc", {
    taskDefinitionArgs: {
        containers: {
            node: {
                image: awsx.ecs.Image.fromDockerBuild("eagle-1", dockerBuild),
            }
        }
    }
})
```

It should be possible to pass the environment as an Output directly as well.  I've opened https://github.com/pulumi/pulumi-docker/pull/257 to improve that.

**Comment by volkanunsal** [comment]
Is there any way you can make the `Output` type truly async, i.e. return a Promise? Typescript compiler can't tell monadic types and type primitives apart, but it can detect when a value is a Promise and tell us that it should be awaited. This seems to be an all too common error in my code, and often hard to detect when it concerns environmental variables. I imagine it wouldn't be convenient to await an Output without top-level awaits, which only became available in 14.8. So there is that. Anyway, maybe the best solution is a linter that can catch this type of error.

**Comment by mikhailshilkov** [comment]
> Is there any way you can make the Output type truly async, i.e. return a Promise? 

See this comment for explanation of why we don't want to do this: https://github.com/pulumi/pulumi/issues/5924#issuecomment-743248303

As the original question has been answered, I'll close the issue
